### PR TITLE
fix/execute raw sql bug

### PR DESCRIPTION
**Pull request by nicholaschen09**
## Summary

fix for https://github.com/InsForge/InsForge/issues/667 where raw executed sql could delete users or other data in the auth schema

fix: adding a regex check in the sanitizeQuery() function to prevent any sql query referencing the auth schema 

## How did you test this change?

1. run these curl commands to create some users: 

```
TOKEN=$(curl -s -X POST "http://localhost:7130/api/auth/admin/sessions" \
  -H "Content-Type: application/json" \
  -d '{"email":"admin@example.com","password":"change-this-password"}' | \
  grep -o '"accessToken":"[^"]*' | cut -d'"' -f4)

curl -X POST "http://localhost:7130/api/auth/users" \
  -H "Authorization: Bearer $TOKEN" \
  -H "Content-Type: application/json" \
  -d '{"email":"testuser1@example.com","password":"test123456","name":"Test User 1"}'

curl -X POST "http://localhost:7130/api/auth/users" \
  -H "Authorization: Bearer $TOKEN" \
  -H "Content-Type: application/json" \
  -d '{"email":"testuser2@example.com","password":"test123456","name":"Test User 2"}'
```
2. verify they exist:

```
curl -X GET "http://localhost:7130/api/auth/users" \
  -H "Authorization: Bearer $TOKEN"
```
3. go to the sql editor in app and try to delete these users by running something like this:

```
DELETE FROM auth.users WHERE email LIKE 'testuser%@example.com'
```
you should see something like this error: 

Screenshot 2025-12-31 at 1 52 19 PM

4. check if they still exist

```
curl -X GET "http://localhost:7130/api/auth/users" \
  -H "Authorization: Bearer $TOKEN"
```

## Summary by CodeRabbit

* **Bug Fixes**
  * Enhanced parser-based validation now blocks additional destructive operations against the authentication schema, emits a warning, and returns a clear forbidden error (403); read-only and allowed actions remain permitted.

* **Tests**
  * Added comprehensive unit tests covering blocked/allowed SQL patterns and edge cases (quoting, whitespace, subqueries, literals) and verifying correct error responses.
  * Adjusted a unit expectation for unknown SQL type handling.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: backend/src/services/database/database-advance.service.ts [hunk 1]
```diff
@@ -89,6 +86,16 @@ export class DatabaseAdvanceService {
       }
     }
 
+    // Block any operations on auth schema (handles double-quoted/unquoted identifiers and whitespace)
```
Comment (tonychang04) [inline backend/src/services/database/database-advance.service.ts]: i think we should still allow selecting things from auth schema. and I'm curious would auth.uid still works.

cc @Fermionic-Lyu  I think until we formally have a role system or a role dedicated for raw sql. We can first do sanitization and limits like this, as it's quicker and easily testable/revertable

**Comment by tonychang04** [comment]
@nicholaschen09  can you address code rabbit's comment?

**Comment by Fermionic-Lyu** [comment]
I think the issue ticket only wants to forbid user deletions from coding agent using raw-sql mcp tool. We still need to allow agent to access the auth schema in different ways, say adding a trigger on auth.users table, so developers can create a profile or user_info table that automatically adds an row whenever a new user registered. Sometimes it's also convenient to ask agent to add some test users as well.

File: backend/src/services/database/database-advance.service.ts [hunk 2]
```diff
@@ -89,10 +92,31 @@ export class DatabaseAdvanceService {
...
+        403,
+        ERROR_CODES.FORBIDDEN
+      );
+    }
+
     return query;
   }
 
   async executeRawSQL(query: string, params: unknown[] = []): Promise<RawSQLResponse> {
+    // Defense-in-depth: Block DELETE operations on auth schema even if sanitizeQuery was bypassed
```
Comment (tonychang04) [inline backend/src/services/database/database-advance.service.ts]: i think we always run sanitize query before execute raw sql so it should be fine, we can remove this code block. maybe in the future we should make sanitize query in execute raw sql

File: backend/src/services/database/database-advance.service.ts [hunk 3]
```diff
@@ -89,10 +92,31 @@ export class DatabaseAdvanceService {
       }
     }
 
+    // Block DELETE operations on auth schema (prevents user deletion via raw SQL)
+    // This is the specific security requirement - prevent coding agents from deleting users
+    const deleteFromAuth = /DELETE\s+.*?\bFROM\s+["']?auth["']?\./i;
```
Comment (tonychang04) [inline backend/src/services/database/database-advance.service.ts]: i only mentioned delete operations, but just curious. can you think of other operations that's dangerous? worth discussing. if not i think this is fine for now as well.

File: backend/tests/unit/database-advance.test.ts [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+        expect(error).toBeInstanceOf(AppError);
+        if (error instanceof AppError) {
+          expect(error.statusCode).toBe(403);
+          expect(error.code).toBe(ERROR_CODES.FORBIDDEN);
+          expect(error.message).toContain('auth schema');
+        }
+      }
+    });
+  });
+
```
Comment (tonychang04) [inline backend/tests/unit/database-advance.test.ts:217]: thx for the tests!

File: backend/src/services/database/database-advance.service.ts (hunk 3 above)
Comment (nicholaschen09) [reply backend/src/services/database/database-advance.service.ts]: hm maybe truncate and drop table?

File: backend/src/services/database/database-advance.service.ts [hunk 5]
```diff
@@ -89,6 +94,39 @@ export class DatabaseAdvanceService {
       }
     }
 
+    // Block DELETE operations on auth schema
+    const deleteFromAuth = /DELETE\s+.*?\bFROM\s+["']?auth["']?\s*\./i;
```
Comment (tonychang04) [inline backend/src/services/database/database-advance.service.ts]: do you think these should be put together or 3 seperate ones? I'm slighly more inclined putting it together if possible

Comment (nicholaschen09) [reply backend/src/services/database/database-advance.service.ts]: okok

File: backend/src/services/database/database-advance.service.ts [hunk 6]
```diff
@@ -89,6 +94,34 @@ export class DatabaseAdvanceService {
...
+        operation = 'DROP';
+        message =
+          'DROP operations on auth schema are not allowed. This would destroy authentication resources and break the system.';
+      }
+
+      logger.warn(`Blocked ${operation} operation on auth schema`, {
+        query: query.substring(0, 100),
+      });
+      throw new AppError(message, 403, ERROR_CODES.FORBIDDEN);
+    }
```
[the remark this answers is not in the document]
Comment (nicholaschen09) [reply backend/src/services/database/database-advance.service.ts]: hm @tonychang04 should i use the sql parser or keep the code like it is currently?

[the remark this answers is not in the document]
Comment (tonychang04) [reply backend/src/services/database/database-advance.service.ts]: code like this is also fine. but we should be able to resolve the false positive code @nicholaschen09 

sql parser is written in a rush and may have edge cases

File: backend/src/services/database/database-advance.service.ts [hunk 7]
```diff
@@ -89,6 +94,86 @@ export class DatabaseAdvanceService {
       }
     }
 
+    const authSchemaBlocked =
```
Comment (tonychang04) [inline backend/src/services/database/database-advance.service.ts]: damn.. this seems very messy.. is this really the best way? can using the parser help? I didn't imagine it would take this much chunk..

File: backend/src/services/database/database-advance.service.ts [hunk 8]
```diff
@@ -89,6 +94,75 @@ export class DatabaseAdvanceService {
...
+      /(?:DELETE\s+(?:FROM\s+["']?auth["']?\s*\.|(?:(?!FROM\s+[^a]).)*?\bFROM\s+["']?auth["']?\s*\.)|TRUNCATE\s+(?:TABLE\s+)?(?:IF\s+EXISTS\s+)?["']?auth["']?\s*\.|DROP\s+.*?["']?auth["']?\s*(?:\.|(?:\s|$|CASCADE)))/i;
+
+    if (authSchemaBlocked.test(query)) {
+      const checkIfInCommentOrString = (matchIndex: number, matchLength: number): boolean => {
```
Comment (tonychang04) [inline backend/src/services/database/database-advance.service.ts]: this would be a bit hard to maintain... and understand and can only use LLM's ofor debugging..  Would using a parser help?

Comment (nicholaschen09) [reply backend/src/services/database/database-advance.service.ts]: ah okok i guess parser would be better then

**Comment by tonychang04** [comment]
would you mind joining our discord for faster review ans response 😂

**Comment by nicholaschen09** [comment]
how do i join the discord can you send link? also is this pr good now?

File: backend/src/utils/sql-parser.ts [hunk 9]
```diff
@@ -92,6 +92,160 @@ function extractChange(stmt: Record<string, unknown>): DatabaseResourceUpdate |
...
+            }
+          }
+        }
+      }
+    }
+
+    return null; // No dangerous operations found
+  } catch (parseError) {
+    // If parsing fails, fall back to regex-based check for safety
+    // This handles edge cases where the parser might fail on complex queries
```
Comment (tonychang04) [inline backend/src/utils/sql-parser.ts]: i would not leave fallback.. if there's an error lets just let  the raw sql pass thorough right?. Ideally we should let the parser work.

else we woulnd't know if it's the parser working or the fallback working for the tests..

File: backend/src/services/database/database-advance.service.ts [hunk 10]
```diff
@@ -69,13 +69,18 @@ export class DatabaseAdvanceService {
...
-   * The API only accesses public schema through PostgREST.
+   * Allows:
+   * - SELECT queries on auth schema (for reading user data)
+   * - INSERT operations on auth schema (for test users)
+   * - CREATE TRIGGER on auth tables (for automatic profile creation, etc.)
+   * - Other DDL operations on auth schema (ALTER TABLE for indexes, etc.)
    */
   sanitizeQuery(query: string, _mode: 'strict' | 'relaxed' = 'strict'): string {
-    // Block database-level operations
+    // Block database-level operations (keep regex for these as they're simple patterns)
```
Comment (tonychang04) [inline backend/src/services/database/database-advance.service.ts]: i don't think we need this comment change? i think this is written by AI?

**Comment by tonychang04** [comment]
help address the code rabbit thanks!

**Comment by nicholaschen09** [comment]
addressed the code rabbit comments!

**Comment by tonychang04** [comment]
https://discord.gg/JJFmcAsS also feel free to join our discord for faster response and reviews!

**Comment by nicholaschen09** [comment]
sounds good! will someone else merge this because i don't have write access
